src/lib.rs

```diff
@@ -9,84 +9,129 @@
+//! }
+//! ```
 //!
 //! The key function is `Rng::gen()`. It is polymorphic and so can be used to
-//! generate any type supporting the [`Uniform`] distribution (i.e. `T` where
-//! `Uniform`: `Distribution<T>`). Type inference means that often a simple call
-//! to `rng.gen()` will suffice, but sometimes an annotation is required, e.g.
+//! generate many types; the [`Uniform`] distribution carries the
+//! implementations. In some cases type annotation is required, e.g.
 //! `rng.gen::<f64>()`.
 //!
-//! See the `distributions` submodule for sampling random numbers from
-//! distributions like normal and exponential.
-//!
-//! # Usage
+//! # Getting random values
 //!
-//! This crate is [on crates.io](https://crates.io/crates/rand) and can be
-//! used by adding `rand` to the dependencies in your project's `Cargo.toml`.
-//!
-//! ```toml
-//! [dependencies]
-//! rand = "0.4"
+//! The most convenient source of randomness is likely `thread_rng()`, which
+//! automatically initialises a fast algorithmic generator on first use per
+//! thread with thread-local storage.
+//! 
+//! If one wants to obtain random data directly from an external source it is
+//! recommended to use `EntropyRng` which manages multiple available sources
+//! or `OsRng` which retrieves random data directly from the OS. It should be
+//! noted that this is significantly slower than using a local generator like
+//! `thread_rng` and potentially much slower if `EntropyRng` must fall back to
+//! `JitterRng` as a source.
+//! 
+//! It is also common to use an algorithmic generator in local memory; for
+//! example `StdRng` or `SmallRng` (the main advantages of the latter are
+//! much smaller state and faster initialisation while its disadvantage is
+//! that it may be easy for an attacker to predict; actual performance varies).
+//! These can be seeded from a parent generator (`SeedableRng::thread_rng`) or
+//! with fresh entropy using `NewRng`:
+//! 
 //! ```
-//!
-//! and this to your crate root:
-//!
-//! ```rust
-//! extern crate rand;
+//! use rand::{SmallRng, NewRng};
+//! let mut rng = SmallRng::new();
 //! ```
-//!
-//! # Thread-local RNG
-//!
-//! There is built-in support for a RNG associated with each thread stored
-//! in thread-local storage. This RNG can be accessed via `thread_rng`, or
-//! used implicitly via `random`. This RNG is normally randomly seeded
-//! from an operating-system source of randomness, e.g. `/dev/urandom` on
-//! Unix systems, and will automatically reseed itself from this source
-//! after generating 32 KiB of random data.
-//!
+//! 
+//! In case one wants specifically to have a reproducible stream of "random"
+//! data (e.g. to procedurally generate a game world), select a named algorithm
+//! (i.e. not `StdRng`/`SmallRng` which may be adjusted in the future), and
+//! use `SeedableRng::from_seed`. Some generators provide additional
+//! constructors suitable for some applications, e.g. `IsaacRng::new_from_u64`.
```

Comment:
Mostly unrelated though: most RNGs now have a `new_unseeded` method. They are a thorn in my eye :smile:. Should we deprecate them with 0.5?

Reply:
I think that was originally the plan but got forgotten? Sure, why not.